### ISSUE TITLE
[Feat]: add `method=similarity` to `api/v2/weights` endpoint

### ISSUE THREAD

**Issue by andrewm4894** [netdata/netdata#14967]
### Problem

Following on from discussion on [this issue](https://github.com/netdata/netdata-cloud/issues/793#issuecomment-1521965435) we would like to add a "metric similarity" capability to the agent. 

At a high level this endpoint would take in a "reference time series" and then perform metric similarity calculation between the "reference time series" and each metric within a specific "candidate set" on the agent. It will return then a similarity score for each metric in the candidate set based on [cosine similarity](https://en.wikipedia.org/wiki/Cosine_similarity) between them.

This will later be exposed in NC to power some sort of "Metric Similarity" features.

### Description

- add a new `method=similarity` to `api/v2/weights`.
- ability to add `reference_data` when making POST to weights endpoint.
- agent will implement recipe similar to [this POC](https://github.com/andrewm4894/netdata-ml-apps/blob/main/pages/Metric_Similarity.py).
- response from weights will be similar the current response with weight values actually just being similarity scores.

### Importance

must have

### Value proposition

1. adds a new foundational capability for netdata to "find similar looking lines" to the "reference time series" (as passed in via `reference_data`)
...

### Proposed implementation

The steps in terms of the algo would be:

```python
# preprocess reference data
reference_data_preprocessed = preprocess(reference_data)
for metric in candidate_set:
    # apply a [min max normalisation](https://en.wikipedia.org/wiki/Feature_scaling#Rescaling_(min-max_normalization)) to each metric so they all on [0,1] scale
    metric_preprocessed = preprocess(metric)
    # calculate the cosine similarity between `reference_data_preprocessed` and `metric_preprocessed`    
    similarity_score = similarity(reference_data, metric_preprocessed)
```

Notes:
- need to make sure that `reference_data` and `metric` are same length - so based on `len(reference_data)` and `after` and `before` will need to make sure same level of aggregation is applied to metric data pulled from agent. e.g. all need to be same number of `points` so we are comparing like with like.
- [Wikipedia](https://en.wikipedia.org/wiki/Cosine_similarity).
- [Good YT video](https://www.youtube.com/watch?v=e9U0QAFbfLI) explaining it (just that instead of feature vector being word counts we would just be using normalized metric values).
- [PyTorch docs](https://pytorch.org/docs/stable/generated/torch.nn.CosineSimilarity.html) on cosine similarity.
- [This SO](https://stackoverflow.com/a/30404185/1919374) seems like an example implementation to start with.
- [OpenCV implementation](https://github.com/opencv/opencv_contrib/blob/master/modules/tracking/src/tracking_by_matching.cpp#L41-L52).
- A really useful [chatgpt4 chat](https://shareg.pt/BpBEwH2) that covers the general problem pretty well, along with various pros and cons of different algos we could use if wanted to get fancier later.

**Comment by andrewm4894** [comment]
@ktsaou @vkalintiris @MrZammler @thiagoftsm based on [this comment](https://github.com/netdata/netdata-cloud/issues/793#issuecomment-1521965435) this issue maps out a potetnail approach to adding a new similarity method to weights endpoint on agent.

**Comment by andrewm4894** [comment]
@vkalintiris fyi

**Comment by andrewm4894** [comment]
@vkalintiris or @ktsaou I'd like to see if we could maybe add this new method to the weights endpoint in next few weeks or so.

This would put in place what we need on the agent side for the metric similarity feature which could be done similar to existing MC with one or two small tweaks. 

https://github.com/netdata/netdata-cloud/issues/793#issuecomment-1521965435
